Add Oklab support

arzg opened the pull request:
Oklab assumes the D65 colour space. In my first attempt at implementing this, the `Oklab` struct didn’t have a `Wp` parameter. This caused all `FromColorUnclamped` derives to not compile, so I added the parameter back with comments warning about how it doesn’t have any effect. Ideally this wouldn’t be needed.

okaneco commented on palette/src/oklab.rs:
```diff
@@ -0,0 +1,792 @@
+    Wp: WhitePoint,
+    T: FloatComponent,
+{
+    fn from_color_unclamped(color: Xyz<Wp, T>) -> Self {
+        let l = from_f64::<T>(M1[0][0]) * color.x
+            + from_f64::<T>(M1[0][1]) * color.y
+            + from_f64::<T>(M1[0][2]) * color.z;
```
There's a module with matrix operations in `src/matrix.rs`.

Same comment for the changes in `xyz.rs`.

Ogeon commented on palette/src/oklab.rs:
```diff
@@ -0,0 +1,786 @@
+}
+
+/// This implementation is **only valid if the XYZ white point is D65**.
+impl<Wp, T> FromColorUnclamped<Xyz<Wp, T>> for Oklab<Wp, T>
```
You _should_ be able to restrict this one too, to use `D65` in place of `Wp` everywhere.

Ogeon commented on palette/src/oklab.rs:
```diff
@@ -0,0 +1,786 @@
+    type Scalar = T;
+
+    fn lighten(&self, amount: T) -> Self {
+        Self::new(self.l + amount * from_f64(1.0), self.a, self.b)
```
You can remove `* from_f64(1.0)`, since it shouldn't have an effect.

arzg replied:
Good catch!

Ogeon reviewed (commented):
This looks really good, white point problems aside, and it seems like a nice color space. I suppose we can go one step further too and add an "Oklch" variant of it too, for completeness' sake.

What were the errors you got from the derive macro when you tried to lock the white point? I would like to have that one figured out, rather than a workaround.

Ogeon commented:
I saw some of your comments over at pale-fire, by the way. Once again, don't hesitate to ask for help or clarifications! This library has a lot of complexity under the hood, as you noticed, and the conversion system is the worst of them all. It's like one of those frontend/backend meme pictures... but with three layers of increasing weirdness.

arzg commented on palette/src/oklab.rs:
```diff
@@ -0,0 +1,786 @@
+#[cfg_attr(feature = "serializing", derive(Serialize, Deserialize))]
+#[palette(
+    palette_internal,
+    white_point = "Wp",
```
I changed `Wp` to `D65` and removed the `Wp` type parameter, and got an error like this

```console
error[E0107]: wrong number of type arguments: expected at most 1, found 2
  --> palette/src/rgb/rgb.rs:46:35
   |
46 | #[derive(Debug, PartialEq, Pixel, FromColorUnclamped, WithAlpha)]
   |                                   ^^^^^^^^^^^^^^^^^^ expected at most 1 type argument
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

at every usage site of the `FromColorUnclamped` derive macro.

arzg replied:
Should I push the ‘more correct’ but broken version that doesn’t have the `Wp` parameter now, or should I wait until we fix all the errors?

Ogeon replied:
My guess is that it expands to a give it a white point it shouldn't have, when implementing it as a part of the default set. There are a few places in the macro code where I make a similar exception for the RGB group of types, so you can program it to have the correct parameters there.

If you push your work in progress, I can help nore and direct you along the way where needed. 🙂 It's always possible to tidy up the commits afterwards.

arzg replied:
I tried my best to fix the issue in eefb47e, but got a different error now about `FromColorUnclamped` not being implemented when converting from an `Xyz` of an arbitrary colour space to an `Oklab`, which only has an implementation for `D65`. Not sure where to go from here 😕

Ogeon replied:
Is this in the other FromColorUnclamped implementations? If not, it's totally expected and desired. It shouldn't switch white point automatically. But I assume that's not the case. We'll have to restrict the white point to D65 all over the place if the target type is Oklab, but it looks like that may require a bit more work than I thought. But I'm leaning towards adding `#nearest_color_path<#white_point, #component>: #from_trait_path<#color_ty>` in the where clause for the Oklab case. I can give it a try and see if it works.

Ogeon replied:
So, I think I managed to fix it. Adding the where clause did the trick, since it basically says that we assume it will work, even when it doesn't. Here's what I arrived at after poking around a bit: https://github.com/Ogeon/palette/commit/6014fc9643693a410f68b777bc26869a6a008103

It would probably be nicer to add a list for special requirement types like this, so the `if` condition doesn't grow out of proportion, but I decided to keep this small.

Ogeon replied:
For example, here's what it generates for `Lab`:

````rust
impl<Wp, T> FromColorUnclamped<Oklab<T>> for Lab<Wp, T> where
    T: FloatComponent,
    Wp: WhitePoint,
    T: FloatComponent,
    Wp: WhitePoint,
    T: FloatComponent,
    Xyz<Wp, T>: FromColorUnclamped<Oklab<T>>, // <--------
````

arzg replied:
That’s pretty cool! How can I rebase your commit onto my branch?

Ogeon replied:
I think you can use `git pull`, the same way I did to get your branch: `git pull git://github.com/Ogeon/palette.git arzg-add-oklab`. Make sure you are in your `add-oklab` when doing it. If that doesn't work, just copy the file content. I don't need my name on it. 😁

arzg replied:
Done, thanks!

Ogeon commented:
Hi! I hope all is well. I just wanted to check in and ask if you are still interested in completing this, or if you would rather hand it over to me (or someone else) to finish it. Either option is fine and all I want to know is how to proceed with this PR. There's no shame in wanting to do other things! 😄

The things I think are left to do are:
 * Rebase on top of the latest `master`, to catch up and solve conflicts.
 * Make it pass the tests by removing the left-over white points in the random generation code.
 * Add an implementation of `bytemuck::Pod` and `bytemuck::Zeroable`. (support for them was added today)
 * Add an `Oklch` color space. This would more or less be a copy of `Lch`, but may need a bit more of poking around in the derive macro.

I will help and guide you if you need it, so don't worry about that.

arzg commented:
Hey, thanks for the kind words! Sorry for disappearing for so long. I’d be interested in finishing this PR.

I just rebased on `master`, but the commit history is quite ugly and broken in some places; would you like me to squash? I’ve also fixed the tests and added the `bytemuck` trait implementations.

Do you think it would be better to keep an `Oklch` implementation in a separate PR, or add it in this one?

Ogeon commented:
Great! Don't worry about being away. It happens. I should also have checked in earlier, but time ended up passing faster than anticipated. But then I thought that it would be a good idea to have this sorted out before I get around to make the next release and all.

Either way, a squash sounds like a good idea! And I think it would be good to have Oklch here in the same PR, mostly just to see that they work well together before merging, because of the special casing in the derive macro.

arzg commented:
I’ve squashed the commits, brought the PR up to date with master and added an Oklch implementation. There are still some failing tests, though. @Ogeon would you mind taking a look?

Ogeon commented:
Great! Thanks! I will have a look later when I get off work. 💼

Ogeon commented on palette/src/xyz.rs:
```diff
@@ -318,6 +318,38 @@ where
+    }
+}
+
+impl<T> FromColorUnclamped<Oklch<T>> for Xyz<D65, T>
+where
+    T: FloatComponent,
+{
+    fn from_color_unclamped(color: Oklch<T>) -> Self {
+        let oklab: Oklab<T> = color.into_color_unclamped();
+        Self::from_color_unclamped(oklab)
+    }
+}
```
Were there any issues with letting the macro derive this automatically?

Ogeon commented:
I think it looks great! I had one question, above, but that should be all. 🎉

Ogeon commented:
Oh, it looks like the random distribution is a bit off for Oklch. I didn't see it earlier because the connection to travis broke when they shut down the .org version. Here's a link until it starts working again https://travis-ci.com/github/Ogeon/palette/jobs/514952866

Ogeon commented on palette/src/oklch.rs:
```diff
@@ -0,0 +1,727 @@
+
+    #[cfg(feature = "random")]
+    test_uniform_distribution! {
+        Oklch<f32> as crate::Oklab {
+            l: (0.0, 1.0),
+            a: (-1.0, 1.0),
+            b: (-1.0, 1.0),
+        },
```
I think I know what's going on with the random numbers. I remember having the same issue before. You need to make the limits for `a` and `b` smaller, to `+/-0.7` (`cos 45°`) to make it look for a square that will fit inside the `OkLch` circle. Otherwise it will not generate any points in the corners and the average density becomes uneven.

Ogeon commented:
I have migrated the CI over to github actions, by the way, so you will need to rebase this branch again. Sorry about that, but the new setup should at least be more responsive.

arzg commented:
@Ogeon any idea why GitHub Actions aren’t running?

Ogeon commented:
I just have to manually approve them for first time contributors...

Ogeon commented:
Alright, all the tests are passing and the code looks good! I think it's time to finally merge this. Thank you for all the work and also for coming back and finishing it!

bors r+

arzg commented:
Thank you ❤️
